sway-core/src/semantic_analysis/context.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{
+    parse_tree::declaration::Purity,
+    semantic_analysis::{
+        ast_node::Mode,
+        declaration::{EnforceTypeArguments, MonomorphizeHelper},
+        Namespace,
+    },
+    type_engine::{insert_type, unify_with_self, TypeId, TypeInfo},
+    CompileResult, CompileWarning, TypeArgument, TypeError,
+};
+use sway_types::{span::Span, Spanned};
+
+/// Contextual state tracked and accumulated throughout type-checking.
+pub struct Context<'ns> {
+    pub(crate) namespace: &'ns mut Namespace,
+
+    // The following set of fields are intentionally private. When a `Context` is passed into a new
+    // node during type checking, these fields should be updated using the `with_*` methods which
+    // provides a new `Context`, ensuring we don't leak our changes into the parent nodes.
+    self_type: TypeId,
+    /// While type-checking an expression, this indicates the expected type.
+    ///
+    /// Assists type inference.
+    type_annotation: TypeId,
+    mode: Mode,
+    // TODO: Should this be removed to `CompileError` and `CompileWarning`?
+    help_text: &'static str,
+    /// Tracks the purity of the context, e.g. whether or not we should be allowed to write to
+    /// storage.
+    purity: Purity,
+}
+
+impl<'ns> Context<'ns> {
```

Comment:
Would you be able to rename this to `TypeSystemContext` or something like that? With the new collection context and stuff going in we will need to have differentiating names.

Reply:
I was hoping we could keep the brevity of `Context` seeing as it's used in so many locations, but if there will also be a `CollectionContext` that sounds fair! Perhaps `TypeCheckContext`?